### PR TITLE
Join: feature dataset

### PR DESCRIPTION
**Pull request by JakobEliasWagner**
# Feature/Cleanup: New Dataset Interface

## Description

This pull request introduces an update to the datat-set handling mechanism.
It aims to leverage PyTorch's Dataset handling features to improve model's performance, scalability, ease of use, and maintainability.

Solves #31 

### Which issue does this PR tackle?

  - Scalability of the data-handling system
  - Efficiency of the data-handling system
  - Ease of Use: Currently new features need to be manually added onto the set
  -  Integration with PyTorch Ecosystem

### How does it solve the problem?

  - Implements a DataSet class that is a child from Pytorch's DataSet.
      - Scalability of the data-handling system (https://pytorch.org/blog/efficient-pytorch-io-library-for-large-datasets-many-files-many-gpus/)
      - Efficiency of the data-handling system (https://pytorch.org/blog/efficient-pytorch-io-library-for-large-datasets-many-files-many-gpus/)
      - Ease of Use: Currently changing the batch size or specific transformations
      - Better integration with PyTorch Ecosystem
  - Separation between the data-loader and data-set allows for greater flexibility.

### How are the changes tested?

  - The previously implemented adjusted unit-tests run without error.

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [x] The PR solves the issue it claims to solve and only this one.
- [x] Changes are tested sufficiently and all tests pass.
- [x] Documentation is complete and well-written.
- [x] Changelog has been updated, if necessary.

### CODE REVIEW

File: notebooks/basics.ipynb [hunk 1]
```diff
@@ -82,11 +95,25 @@
     "# Target function (the derivative of u)\n",
     "v = lambda y: torch.cos(y)\n",
```
Comment (samuelburbulla) [inline notebooks/basics.ipynb:96]: Does it make sense to call the callables `u_func` and `v_func` as we will re-use `u` and `v` for tensors below?

File: notebooks/basics.ipynb [hunk 2]
```diff
@@ -131,16 +160,12 @@
     }
    ],
    "source": [
-    "num_sensors = 32\n",
```
Comment (samuelburbulla) [inline notebooks/basics.ipynb:134]: As the number of sensors now is set further above, we should remove the mentioning in the description above this cell.

File: src/continuity/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        for dim, val in sample.items():
+            if dim in self.transform:
+                sample[dim] = self.transform[dim](val)
+
+        return sample["x"], sample["u"], sample["y"], sample["v"]
+
+
+class SelfSupervisedOperatorDataset(OperatorDataset):
+    """
+    A `SelfSupervisedOperatorDataset` is a data set that exports batches of observations
```
Comment (samuelburbulla) [inline src/continuity/data/dataset.py:115]: no batches any more

File: src/continuity/data/dataset.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    positions, `u` the sensor values, and `y = x_i` and `v = u_i` are
+    the label's coordinate its value for all `i`.
+
+    Args:
+        x: Sensor positions of shape (num_observations, num_sensors, coordinate_dim)
+        u: Sensor values of shape (num_observations, num_sensors, num_channels)
+    """
+
+    def __init__(self, x: torch.Tensor, u: torch.Tensor):
+        self.num_observations = u.shape[0]
```
Comment (samuelburbulla) [inline src/continuity/data/dataset.py:129]: Can we use DatasetShape here?

File: src/continuity/operators/operator.py [hunk 5]
```diff
@@ -78,8 +81,8 @@ def fit(
             loss_train = 0
 
             start = time()
-            for sample in dataset:
-                x, u, y, v = sample
+            for x, u, y, v in data_loader:
+                x, u, y, v = x.to(device), u.to(device), y.to(device), v.to(device)
```
Comment (samuelburbulla) [inline src/continuity/operators/operator.py:85]: We shouldn't perform device management here, as we would also have to move the operator itself to the right device (I get a mismatch at my Mac at many places right now).

Let's assume (for now) that per default everything is on the cpu.

File: tests/operator/test_deep_o_net.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import torch
```
Comment (samuelburbulla) [inline tests/operator/test_deep_o_net.py:1]: Please rename to `test_deeponet.py`

File: tests/test_convolution.py [hunk 7]
```diff
@@ -1,27 +1,29 @@
...
+from continuity.data import Sine, device
 from continuity.operators import ContinuousConvolution
-from continuity.plotting import plot
 
 # Set random seed
 torch.manual_seed(0)
 
 
 def test_convolution():
+    torch.set_default_dtype(torch.float64)
```
Comment (samuelburbulla) [inline tests/test_convolution.py:12]: MPS does not support float64. Please use float32 instead.

File: tests/test_convolution.py [hunk 8]
```diff
@@ -1,27 +1,29 @@
...
-    x, u = dataset.x[0], dataset.u[0]
+    dataloader = DataLoader(dataset, batch_size=1, shuffle=True)
+    x, u, _, _ = next(iter(dataloader))
 
     # Kernel
     def dirac(x, y):
         dist = ((x - y) ** 2).sum(dim=-1)
         zero = torch.zeros(1)
-        return torch.isclose(dist, zero).to(torch.float32)
+        return torch.isclose(dist, zero).to(torch.float64)
```
Comment (samuelburbulla) [inline tests/test_convolution.py:26]:
```suggestion
        return torch.isclose(dist, zero).to(torch.float32)
```

File: tests/test_convolution.py [hunk 9]
```diff
@@ -31,18 +33,16 @@ def dirac(x, y):
     )
 
     # Create tensors
-    y = torch.linspace(-1, 1, num_evals).unsqueeze(-1)
+    y = torch.linspace(-1, 1, num_evals).reshape(1, -1, 1).to(device)
```
Comment (samuelburbulla) [inline tests/test_convolution.py:36]: Should not move to device here (I get a device mismatch)

File: src/continuity/data/sine.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        x = torch.linspace(-1, 1, self.num_sensors).reshape(-1, 1)
+
+        if self.size == 1:
+            w = 1
+        else:
+            w = 1 + i / (self.size - 1)
+
+        u = torch.sin(w * torch.pi * x)
+
+        return torch.tensor(x), torch.tensor(u)
```
Comment (samuelburbulla) [inline src/continuity/data/sine.py]:
```suggestion
        return x, u
```

File: src/continuity/data/shape.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+
+@dataclass
+class TensorShape:
+    num: int  # number of separate instances
+    dim: int  # dimensionality of a single instance (needs to be flat)
+
+
+@dataclass
+class DatasetShape:
```
Comment (samuelburbulla) [inline src/continuity/data/shape.py:11]: Can we rename this to `DatasetShapes`?

File: src/continuity/operators/deeponet.py [hunk 12]
```diff
@@ -32,9 +28,7 @@ class DeepONet(Operator):
 
     def __init__(
         self,
-        num_sensors: int,
-        coordinate_dim: int = 1,
-        num_channels: int = 1,
+        dataset_shape: DatasetShape,
```
Comment (samuelburbulla) [inline src/continuity/operators/deeponet.py:31]: I would prefer `shapes` for shortness (see also comment on `DatasetShapes`).

Then, e.g., line below becomes:
`self.dot_dim = shapes.v.num * shapes.v.dim * basis_functions`

File: notebooks/selfsupervised.ipynb [hunk 13]
```diff
@@ -87,12 +101,18 @@
       "Dataloader contains 32 batches.\n"
      ]
     },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/jakob/master_workspace/code/continuity/src/continuity/data/sine.py:69: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
```
Comment (samuelburbulla) [inline notebooks/selfsupervised.ipynb]: Should not include any warning (particularly containing a personal path).
The warning will be fixed by another comment below.

File: notebooks/superresolution.ipynb [hunk 14]
```diff
@@ -84,25 +97,36 @@
...
+      "\u001B[0;31mFileNotFoundError\u001B[0m: [Errno 2] No such file or directory: '../data/flame/train.csv'"
```
Comment (samuelburbulla) [inline notebooks/superresolution.ipynb]: We should generate the correct output for this notebook. If you don't have the flame data, I can run it and push in the end.

**Review by samuelburbulla — CHANGES_REQUESTED** [review]

File: notebooks/basics.ipynb [hunk 15]
```diff
@@ -131,16 +160,12 @@
...
     "# Define operator\n",
-    "operator = DeepONet(num_sensors)\n",
-    "\n",
-    "# Some sensor positions\n",
-    "x = torch.linspace(-L, L, num_sensors)\n",
+    "operator = DeepONet(dataset_shape=dataset.shape)\n",
     "\n",
     "# Plot sensors\n",
     "plt.plot(x, u(x), \"go\", label=\"Sensors\")\n",
+    "plt.plot(y, u(y), \"g--\", alpha=0.5, label=\"Input Function\")\n",
```
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb:170]: Maybe introducing a third variable from the domain of x and y would be beneficial here. With a third variable it would be clearer whether continuous representations of these functions are plotted/used and when a discretization of the domain is used.

File: notebooks/basics.ipynb [hunk 16]
```diff
@@ -82,11 +95,25 @@
...
+    "num_sensors = 32\n",
+    "num_evaluations = 100\n",
+    "x = torch.linspace(-L, L, num_sensors)\n",
+    "y = torch.linspace(-L, L, num_evaluations)  # here the domain of the derivative is the same as the domain of the function\n",
+    "\n",
+    "# This dataset contains only a single sample (first dimension of all tensors)\n",
+    "dataset = OperatorDataset(\n",
+    "    x=x.reshape(1, -1, 1),\n",
+    "    u=u(x).reshape(1, -1, 1),\n",
+    "    y=y.reshape(1, 1, -1),\n",
```
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb]: Resolved in [e0af0d0](https://github.com/aai-institute/continuity/pull/35/commits/e0af0d03167c41676e09c4a76e4b6049a26a5c90)

File: notebooks/basics.ipynb (hunk 15 above)
[the remark this answers is not in the document]
Comment (samuelburbulla) [reply notebooks/basics.ipynb:170]: I think `x` is perfectly fine here as it is the input for `u`. In this example we have `x=y`, but we should be consistent when we call the corresponding functions. We never pass a continuous representation of u's input domain, do we?

File: notebooks/basics.ipynb [hunk 17]
```diff
@@ -222,22 +236,27 @@
...
+    }
+   },
    "outputs": [
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "Model parameters: 7984\n",
-      "Epoch 1000:  loss/train = 1.4092e-04  (0.00 s/epoch)\n"
+      "Model parameters: 63424\n",
```
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb]: This was due to an error in the dot-product space. Fixed in [d983e3c](https://github.com/aai-institute/continuity/pull/35/commits/d983e3c51d6e6efd5677fe966725e7cb7203793f)

File: notebooks/basics.ipynb [hunk 18]
```diff
@@ -82,11 +95,25 @@
...
+    "dataset = OperatorDataset(\n",
+    "    x=x.reshape(1, -1, 1),\n",
+    "    u=u(x).reshape(1, -1, 1),\n",
+    "    y=y.reshape(1, 1, -1),\n",
+    "    v=v(y).reshape(1, 1, -1)\n",
+    ")\n",
+    "data_loader = DataLoader(dataset)\n",
+    "\n",
+    "# Plot input and target function\n",
+    "plt.plot(y, u(y), \"g-\", label=\"Input $u$\")\n",
```
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb]: change in in [24a242c](https://github.com/aai-institute/continuity/pull/35/commits/24a242cb3b5636d1f7f346054fc94895772295f0)

File: notebooks/basics.ipynb (hunk 1 above)
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb]: resolved in [24a242c](https://github.com/aai-institute/continuity/pull/35/commits/24a242cb3b5636d1f7f346054fc94895772295f0)

File: notebooks/basics.ipynb (hunk 2 above)
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb:134]: resolved in [24a242c](https://github.com/aai-institute/continuity/pull/35/commits/24a242cb3b5636d1f7f346054fc94895772295f0)

File: tests/operator/test_deep_o_net.py (hunk 6 above)
Comment (JakobEliasWagner) [reply tests/operator/test_deep_o_net.py]: resolved in [6325b4a](https://github.com/aai-institute/continuity/pull/35/commits/6325b4aef6ad772ab51d0dc72a87dd94412439cc)

File: tests/test_convolution.py (hunk 7 above)
Comment (JakobEliasWagner) [reply tests/test_convolution.py]: Added `torch.get_default_dtype()` instead in [6325b4a](https://github.com/aai-institute/continuity/pull/35/commits/6325b4aef6ad772ab51d0dc72a87dd94412439cc).

File: tests/test_convolution.py (hunk 8 above)
Comment (JakobEliasWagner) [reply tests/test_convolution.py]: Added `torch.get_default_dtype()` instead in [6325b4a](https://github.com/aai-institute/continuity/pull/35/commits/6325b4aef6ad772ab51d0dc72a87dd94412439cc).

File: tests/test_convolution.py (hunk 9 above)
Comment (JakobEliasWagner) [reply tests/test_convolution.py:36]: moved the operator to the device as well in [6325b4a](https://github.com/aai-institute/continuity/pull/35/commits/6325b4aef6ad772ab51d0dc72a87dd94412439cc).

File: src/continuity/operators/deeponet.py (hunk 12 above)
Comment (JakobEliasWagner) [reply src/continuity/operators/deeponet.py]: adapted to this in [c3be4ea](https://github.com/aai-institute/continuity/pull/35/commits/c3be4eaed9400872c0f2ac7d6784d33ccff39540).

File: src/continuity/data/dataset.py (hunk 4 above)
Comment (JakobEliasWagner) [reply src/continuity/data/dataset.py:129]: No we can't as currently DatasetShape describes individual observations. Adjusting `DatasetShapes` to this would be beneficial.

File: src/continuity/data/shape.py (hunk 11 above)
Comment (JakobEliasWagner) [reply src/continuity/data/shape.py]: adapted in [42497cd](https://github.com/aai-institute/continuity/pull/35/commits/42497cd02ee2bf3602da8a6bd0721e0a56b2bc66).

File: src/continuity/data/dataset.py (hunk 4 above)
Comment (JakobEliasWagner) [reply src/continuity/data/dataset.py:130]: adapted in [42497cd](https://github.com/aai-institute/continuity/pull/35/commits/42497cd02ee2bf3602da8a6bd0721e0a56b2bc66).

File: notebooks/selfsupervised.ipynb (hunk 13 above)
Comment (JakobEliasWagner) [reply notebooks/selfsupervised.ipynb]: fixed in [92d8a65](https://github.com/aai-institute/continuity/pull/35/commits/92d8a652f9a789a262b93dd684577cf3fbca9d21).

File: notebooks/superresolution.ipynb (hunk 14 above)
Comment (JakobEliasWagner) [reply notebooks/superresolution.ipynb]: No I do not have it.

File: src/continuity/data/dataset.py (hunk 3 above)
Comment (JakobEliasWagner) [reply src/continuity/data/dataset.py]: fixed in [92d8a65](https://github.com/aai-institute/continuity/pull/35/commits/92d8a652f9a789a262b93dd684577cf3fbca9d21).

File: src/continuity/operators/operator.py (hunk 5 above)
Comment (JakobEliasWagner) [reply src/continuity/operators/operator.py:88]: Moved the operator to the device before trainig in [870da52](https://github.com/aai-institute/continuity/pull/35/commits/870da52b20fbe8e80be8e8e14b6c52ab2d8fc865) .

File: notebooks/basics.ipynb (hunk 15 above)
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply notebooks/basics.ipynb]: Saw your suggestion too late. Introduced xi to the notebook. Is this okay?

**Comment by JakobEliasWagner** [comment]
@samuelburbulla added your suggested changes to the PR. Need some clarification on whether you think it is okay to use xi in basics.ipynb. Furthermore, I do not have access to the flame dataset.

File: notebooks/superresolution.ipynb (hunk 14 above)
Comment (samuelburbulla) [reply notebooks/superresolution.ipynb]: Pushed the output.

**Review by samuelburbulla — APPROVED** [review]
